src/v/cluster/archival_metadata_stm.cc

```diff
@@ -1491,44 +1491,42 @@ void archival_metadata_stm::apply_reset_scrubbing_metadata() {
     _manifest->reset_scrubbing_metadata();
 }
 
-std::vector<cloud_storage::partition_manifest::lw_segment_meta>
+fragmented_vector<cloud_storage::partition_manifest::lw_segment_meta>
 archival_metadata_stm::get_segments_to_cleanup() const {
     // Include replaced segments to the backlog
     using lw_segment_meta = cloud_storage::partition_manifest::lw_segment_meta;
-    std::vector<lw_segment_meta> backlog = _manifest->lw_replaced_segments();
+    const fragmented_vector<lw_segment_meta> source_backlog
+      = _manifest->lw_replaced_segments();
 
     // Make sure that 'replaced' list doesn't have any references to active
     // segments. This is a protection from the data loss. This should not
     // happen, but protects us from data loss in cases where bugs elsewhere.
-    auto backlog_size = backlog.size();
-    backlog.erase(
-      std::remove_if(
-        backlog.begin(),
-        backlog.end(),
-        [this](const lw_segment_meta& m) {
-            auto it = _manifest->find(m.base_offset);
-            if (it == _manifest->end()) {
-                return false;
-            }
-            auto m_name = _manifest->generate_remote_segment_name(
-              cloud_storage::partition_manifest::lw_segment_meta::convert(m));
-            auto s_name = _manifest->generate_remote_segment_name(*it);
-            // The segment will have the same path as the one we have in
-            // manifest in S3 so if we will delete it the data will be lost.
-            if (m_name == s_name) {
-                vlog(
-                  _logger.error,
-                  "The replaced segment name {} collides with the segment "
-                  "{} "
-                  "in the manifest. It will be removed to prevent the data "
-                  "loss.",
-                  m_name,
-                  s_name);
-                return true;
-            }
-            return false;
-        }),
-      backlog.end());
+    const auto backlog_size = source_backlog.size();
+    fragmented_vector<lw_segment_meta> backlog;
+    for (const auto& m : source_backlog) {
+        auto it = _manifest->find(m.base_offset);
+        if (it == _manifest->end()) {
+            backlog.push_back(m);
+            continue;
+        }
+        auto m_name = _manifest->generate_remote_segment_name(
+          cloud_storage::partition_manifest::lw_segment_meta::convert(m));
+        auto s_name = _manifest->generate_remote_segment_name(*it);
+        // The segment will have the same path as the one we have in
+        // manifest in S3 so if we will delete it the data will be lost.
+        if (m_name == s_name) {
+            vlog(
+              _logger.error,
+              "The replaced segment name {} collides with the segment "
+              "{} "
+              "in the manifest. It will be removed to prevent the data "
+              "loss.",
+              m_name,
+              s_name);
+            continue;
+        }
+        backlog.push_back(m);
+    }
```

Comment:
nit: `copy_if` with a predicate seems like it would be easier to read as the predicate just has to evaluate to true or false.

Reply:
https://github.com/redpanda-data/redpanda/pull/16189